src/main/java/com/artipie/http/TimeoutSlice.java

```diff
@@ -0,0 +1,213 @@
+
+        /**
+         * Scheduler.
+         */
+        private final ScheduledExecutorService exec;
+
+        /**
+         * Ctor.
+         * @param origin Origin response
+         * @param duration Timeout duration
+         * @param exec Executor service
+         */
+        TimeoutResponse(final Response origin, final Duration duration,
+            final ScheduledExecutorService exec) {
+            this.origin = origin;
+            this.duration = duration;
+            this.exec = exec;
+        }
+
+        @Override
+        public CompletionStage<Void> send(final Connection connection) {
+            final TimeoutConnection tcon = new TimeoutConnection(connection, this.exec);
+            return CompletableFuture.anyOf(
+                this.origin.send(tcon).toCompletableFuture(),
+                tcon.timeout(this.duration)
+            ).thenApply(ignore -> null);
+        }
+    }
+
+    /**
+     * Connection with timeout support.
+     * @since 0.10
+     */
+    private static final class TimeoutConnection implements Connection {
+
+        /**
+         * Origin connection.
+         */
+        private final Connection origin;
+
+        /**
+         * Scheduler.
+         */
+        private final ScheduledExecutorService exec;
+
+        /**
+         * Completed flag.
+         */
+        private final AtomicBoolean completed;
+
+        /**
+         * Ctor.
+         * @param origin Origin connection
+         * @param exec Scheduler
+         */
+        TimeoutConnection(final Connection origin, final ScheduledExecutorService exec) {
+            this.origin = origin;
+            this.exec = exec;
+            this.completed = new AtomicBoolean();
+        }
+
+        @Override
+        public CompletionStage<Void> accept(final RsStatus status, final Headers headers,
+            final Publisher<ByteBuffer> body) {
+            final CompletionStage<Void> res;
+            if (this.completed.compareAndSet(false, true)) {
+                res = this.origin.accept(status, headers, body);
+            } else {
+                res = CompletableFuture.completedFuture(null);
+            }
+            return res;
+        }
+
+        /**
+         * Start timeout task.
+         * @param duration Timeout duration
+         * @return Task future
+         */
+        CompletableFuture<?> timeout(final Duration duration) {
+            final CompletableFuture<?> future = new CompletableFuture<>();
+            this.exec.schedule(
+                () -> {
+                    if (this.completed.compareAndSet(false, true)) {
+                        this.origin.accept(
+                            RsStatus.UNAVAILABLE, Headers.EMPTY,
+                            new Content.From(
+                                "Request cancelled by timeout".getBytes(StandardCharsets.UTF_8)
+                            )
+                        );
+                        future.complete(null);
```

Comment:
@g4s8 `future` should complete only after `this.origin.accept` has been completed, please move `future.complete` invocation to `.thenRun()`